### PR TITLE
Vcx updates zkp IS-1174

### PR DESCRIPTION
**Pull request by ianco**
Updates to VCX for ZKP

Add predicate support to proof requests and proofs for VCX

### CODE REVIEW

**Comment by jovfer** [comment]
@glowkey @Artemkaaas could you review please?

**Comment by ianco** [comment]
> These changes look good. How much effort would it be to add an additional test for the negative condition?

I'll look into adding the test

**Comment by ianco** [comment]
Added unit tests

**Comment by ianco** [comment]
Added a fix for an edge case I found - see unit tests tests_libindy_prover_get_credentials (fetch with mal-formed json) vs test_retrieve_credentials_emtpy (fetch with valid json but returns no requested attributes).

File: vcx/wrappers/python3/vcx/api/proof.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def proof_state(self, x):
         self._proof_state = x
 
     @staticmethod
-    async def create(source_id: str, name: str, requested_attrs: list, revocation_interval: dict):
```
Comment (Artemkaaas) [inline vcx/wrappers/python3/vcx/api/proof.py]: What about others vcx wrapper? Do we need updates there?

File: vcx/libvcx/src/disclosed_proof.rs [hunk 2]
```diff
@@ -300,21 +310,35 @@ impl DisclosedProof {
         Ok(rtn.to_string())
     }
 
-    fn build_requested_credentials_json(&self, credentials_identifiers: &Vec<CredInfo>, self_attested_attrs: &str) -> VcxResult<String> {
+    fn build_requested_credentials_json(&self, 
+                                        credentials_identifiers: &Vec<CredInfo>, 
+                                        self_attested_attrs: &str,
+                                        proof_req: &ProofRequestData) -> VcxResult<String> {
         let mut rtn: Value = json!({
```
[the remark this answers is not in the document]
Comment (ianco) [reply vcx/libvcx/src/disclosed_proof.rs]: Probably, but the use of the serde Value is not part of the change I made (this part of the code is a year old) and is a bigger change than I was planning to get into adding the predicate support ...

File: vcx/libvcx/src/disclosed_proof.rs [hunk 3]
```diff
@@ -829,7 +856,22 @@ mod tests {
...
+                "height_1": {
+                    "name": "height_1",
+                    "non_revoked":  {"from": 123, "to": 456}
+                },
+                "zip_2": { "name": "zip_2" }
+            },
+            "requested_predicates": {},
+            "non_revoked": {"from": 098, "to": 123}
+        });
+        let proof_req: ProofRequestData = serde_json::from_value(proof_req).unwrap();
```
[the remark this answers is not in the document]
Comment (ianco) [reply vcx/libvcx/src/disclosed_proof.rs]: This is how ProofRequestData is handled consistently in all unit tests.  The nonce, name and version aren't explicitly "pub".

File: vcx/wrappers/python3/vcx/api/proof.py (hunk 1 above)
Comment (ianco) [reply vcx/wrappers/python3/vcx/api/proof.py]: Good catch, I'm not familiar with any of the other wrappers (Java, ios, node) but they would need a similar change.

I don't have a local environment that I can test changes to any of these wrappers.

File: vcx/libvcx/src/disclosed_proof.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Artemkaaas) [reply vcx/libvcx/src/disclosed_proof.rs]: ok. agree. I believe we can skip it
